text/3502-cargo-script.md

```diff
@@ -0,0 +1,1202 @@
+- All crates will be available
+- Local resources are available
+
+**One-Off Utilities:**
+
+It is fairly trivial to create a bunch of single-file bash or python scripts
+into a directory and add it to the path.
+Compare this to rust where
+- `cargo new` each of the "scripts" into individual directories
+- Create wrappers for each so you can access it in your path, passing `--manifest-path` to `cargo run`
+
+**Non-Goals:**
+
+With that said, this doesn't have to completely handle every use case for
+Collaboration, Interoperability, Prototuping, or One-off Utilities.
+Users can always scale up to normal packages with an explicit `Cargo.toml` file.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+### Creating a New Package
+
+*(Adapted from [the cargo book](https://doc.rust-lang.org/cargo/guide/creating-a-new-project.html))*
+
+To start a new [package][def-package] with Cargo, create a file named `hello_world.rs`:
+```rust
+#!/usr/bin/env cargo
+
+fn main() {
+    println!("Hello, world!");
+}
+```
+
+Let's run it
+```console
+$ chmod +x hello_world.rs
+$ ./hello_world.rs
+Hello, world!
+```
+
+### Dependencies
+
+*(Adapted from [the cargo book](https://doc.rust-lang.org/cargo/guide/dependencies.html))*
+
+[crates.io] is the Rust community's central [*package registry*][def-package-registry]
+that serves as a location to discover and download
+[packages][def-package]. `cargo` is configured to use it by default to find
+requested packages.
+
+#### Adding a dependency
+
+To depend on a library hosted on [crates.io], you modify `hello_world.rs`:
+````rust
+#!/usr/bin/env cargo
+```cargo
+[dependencies]
+time = "0.1.12"
+```
+
+fn main() {
+    println!("Hello, world!");
+}
+````
+
+The data inside the `cargo` frontmatter is called a
+[***manifest***][def-manifest], and it contains all of the metadata that Cargo
+needs to compile your package.
+This is written in the [TOML] format (pronounced /tɑməl/).
+
+`time = "0.1.12"` is the name of the [crate][def-crate] and a [SemVer] version
+requirement. The [specifying
+dependencies](https://doc.rust-lang.org/cargo/guide/../reference/specifying-dependencies.html) docs have more
+information about the options you have here.
+
+If we also wanted to add a dependency on the `regex` crate, we would not need
+to add `[dependencies]` for each crate listed. Here's what your whole
+`hello_world.rs` file would look like with dependencies on the `time` and `regex`
+crates:
+
+````rust
+#!/usr/bin/env cargo
+```cargo
+[dependencies]
+time = "0.1.12"
+regex = "0.1.41"
+```
+
+fn main() {
+    let re = Regex::new(r"^\d{4}-\d{2}-\d{2}$").unwrap();
+    println!("Did our date match? {}", re.is_match("2014-01-01"));
+}
+````
+
+You can then re-run this and Cargo will fetch the new dependencies and all of their dependencies.  You can see this by passing in `--verbose`:
+```console
+$ cargo eval --verbose ./hello_world.rs
```

Comment:
`cargo eval` was an earlier proposal, AFAICT? Looks like this example needs to be updated.